shell/browser/ui/file_dialog_linux.cc

```diff
@@ -0,0 +1,260 @@
+// Copyright (c) 2024 Microsoft, Gmbh.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include <memory>
+#include <string>
+
+#include "base/files/file_util.h"
+#include "base/functional/bind.h"
+#include "base/functional/callback.h"
+#include "base/memory/raw_ptr.h"
+#include "base/memory/raw_ptr_exclusion.h"
+#include "base/run_loop.h"
+#include "base/strings/string_util.h"
+#include "base/strings/utf_string_conversions.h"
+#include "shell/browser/javascript_environment.h"
+#include "shell/browser/native_window_views.h"
+#include "shell/browser/ui/file_dialog.h"
+#include "shell/browser/ui/select_file_policy.h"
+#include "shell/common/gin_converters/callback_converter.h"
+#include "shell/common/gin_converters/file_path_converter.h"
+#include "ui/shell_dialogs/select_file_dialog.h"
+#include "ui/shell_dialogs/selected_file_info.h"
+
+namespace file_dialog {
+
+DialogSettings::DialogSettings() = default;
+DialogSettings::DialogSettings(const DialogSettings&) = default;
+DialogSettings::~DialogSettings() = default;
+
+namespace {
+
+ui::SelectFileDialog::Type GetDialogType(int properties) {
+  if (properties & OPEN_DIALOG_OPEN_FILE) {
+    return ui::SelectFileDialog::SELECT_OPEN_FILE;
+  } else if (properties & OPEN_DIALOG_OPEN_DIRECTORY) {
+    return ui::SelectFileDialog::SELECT_FOLDER;
+  } else if (properties & OPEN_DIALOG_MULTI_SELECTIONS) {
+    return ui::SelectFileDialog::SELECT_OPEN_MULTI_FILE;
+  }
+
+  return ui::SelectFileDialog::SELECT_OPEN_FILE;
+}
+
+// [save_label, show_overwrite_confirmation, show_hidden, ]
+std::tuple<std::string, bool, bool> GetSettingsTuple(
+    const DialogSettings& settings) {
+  bool show_overwrite_confirmation =
+      settings.properties & SAVE_DIALOG_SHOW_OVERWRITE_CONFIRMATION;
+  bool show_hidden = settings.properties & SAVE_DIALOG_SHOW_HIDDEN_FILES;
+
+  return std::make_tuple(settings.button_label, show_overwrite_confirmation,
+                         show_hidden);
+}
```

Comment:
Better to use `gtk::ExtraSettings` here than a tuple, since that way if `ExtraSettings` ever changes in `gtk/` then this code won't break.

For the same reason, I'd name this `GetExtraSettings()` to remove the reference to "tuple" :smile_cat: 

```c++
gtk::ExtraSettings GetExtraSettings(const DialogSettings& settings) {
  gtk::ExtraSettings extra;
  extra.button_label = settings.button_label;
  extra.show_overwrite_confirmation = settings.properties & SAVE_DIALOG_SHOW_OVERWRITE_CONFIRMATION;
  extra.show_hidden = settings.properties & SAVE_DIALOG_SHOW_HIDDEN_FILES;
  return extra;
}
```

^ this snippet assumes the other suggestion of making `ExtraSettings` a struct, but IMO the function name and return type change is worthwhile whether `ExtraSettings` is a struct or tuple